server/notebooks/views.py

```diff
@@ -216,3 +217,47 @@ def tryit_view(request):
             "eval_frame_origin": EVAL_FRAME_ORIGIN,
         },
     )
+
+
+@require_http_methods(["POST"])
+@csrf_exempt
+def from_template_view(request):
+    iomd = urllib.parse.unquote(request.POST.get("iomd", ""))
+    if not iomd:
+        return HttpResponseBadRequest("Must specify iomd template")
+
+    title = request.POST.get("title")
+
+    # NOTE: We can not sadly not use files (request.FILES) attached to forms, as
+    # it is not possible to attach those programatically from the client side via
+    # JavaScript
+    files = []
+    for filename in [
+        k.replace("file-", "", 1) for k in request.POST.keys() if k.startswith("file-")
+    ]:
+        if not filename:
+            return HttpResponseBadRequest("Filename must have at least one character")
+        file_content = request.POST[f"file-{filename}"]
+        actual_file_size = len(file_content) * 3 / 4  # base64 size = 4 * actual / 3
+        if actual_file_size > MAX_FILE_SIZE:
+            return HttpResponseBadRequest(
+                f"File {filename} exceeds maximum file size {MAX_FILE_SIZE}"
+            )
```

Comment:
I'm really worried that this could be used to DOS the server, I would feel much better if we would be able to handle this validation via a Django FileField that has been well tested and has a good approach to streaming files to /tmp if needed.

Reply:
We should be respecting the existing limit on the size of uploads, which limits the total size of the form to 2.5 MB:

https://docs.djangoproject.com/en/3.0/ref/settings/#data-upload-max-memory-size

I'm not sure if there's really any more risk here than there would be of the user spamming us through any of the existing mechanisms provided by the server. Note that this method doesn't actually *create* anything on the server -- it just echoes the content of the files back to the client.

Reply:
Right, I meant that `len(file_content)` is risky in terms of memory usage when accepting any content without rate limiting. The big advantage of FileField is that it uses file system calls to determine the size of a temporarily stored file, which is a lot less problematic.

Reply:
Gotcha, I would certainly prefer to use FileField as well were it possible. Unfortunately it doesn't seem like it is given the above. :(

Reply:
Turns out it was possible to use Django's file handling after all by using some client side tricks to add the files to the form.